### ISSUE TITLE
iframe embedding a pad

### ISSUE THREAD

**Issue by joshr** [ether/etherpad-lite#4031]
Hi all, new here and a huge fan of etherpad for a very long time.

i'm building a site that is made up of etherpad pages in iframes but i'm having problems. i'm essentially using the embed code from the embed code button in the edit bar, but i'm making the iframe full page sized.

i've been chatting to John, who's gone above and beyond helping me get this far (i'm not a coder really!)

in safari and brave browsers users aren't able to load the pads in iframes at all due to (i think) cross domain cookies being blocked by the browser, etherpad is blocking being embedded in iframes altogether, or combo of those two things. some reports of this happening in incognito mode in other browsers too.

here's an example page https://reorganise.org/etherpad-page/ to try in safari (this is a wordpress generated page)
my etherpad install is at https://etherpad.xyz (digital ocean, ubuntu, nginx)

i'm on the latest etherpad-lite, and i'm running the latest versions of: comments_page, embedded_hyperlinks2, headings2, set_title_on_pad, spellcheck, table_of_contents

john and i have been thinking to solve this with helmet, but it occurred to me this is an out-of-the-box feature that doesn't/shouldn't rely on a plugin, since iframe embedding is in the vanilla install.

many thanks in advance for advice, fixes, and diagnosis!

**Comment by aspiers** [comment]
Don't suppose https://github.com/vitr/safari-cookie-in-iframe is any use here?

**Comment by JohnMcLear** [comment]
I wonder if cors might do the trick.  https://www.npmjs.com/package/cors & https://github.com/helmetjs/helmet/issues/79

**Comment by JohnMcLear** [comment]
I created a test scenario for this just to show it's not Etherpad that's broken here: 

Test @ https://admissivefalcon.htmlpasta.com/

All I'm doing is putting cookie set and get from mozilla MDN docs into the iframe HTML.

```
<!DOCTYPE html>
<html>
<body>
<button onclick="alertCookie()">Show cookies</button>
<script type="text/javascript">
document.cookie = "name=oeschger";
document.cookie = "favorite_food=tripe";
function alertCookie() {
  alert(document.cookie);
}
</script>
</body>
</html>
```

And this fails in Brave/Safari

**Comment by JohnMcLear** [comment]
https://gist.github.com/iansltx/18caf551baaa60b79206  I think this would need a hack for Safari, which frankly I'm not willing to do.  

The TLDR is that Safari/Brave need a cookie set from the domain providing Etherpad prior to visiting a pad..   @joshr needs to be able to direct his users to Wordpress so the first time the user ever visits a pad he has probably never been on either Wordpress or Etherpad before so redirecting is going to be a mess.

@joshr your best option here is to run etherpad at https://etherpad.reorganise.org 

If someone is looking for a solution for running cookies across domains you can do this:...

When a SAFARI ONLY user visits your site: http://foo.com and they click a link to open a pad, set that link to something like..  http://etherpaddomain.com/redirectWithCookie?redirectTo=someLocation -- Warning: Do not implement this for any one other than those using a crippled browser.  The redirect logic could be a plugin, I'm not willing to write this though.

redirectWithCookie will need to set a cookie for etherpaddomain.com then redirect back to a url that will have etherpad embedded.

_TFLDR; Safari sucks._   Tell your users to use Firefox.

**Comment by JohnMcLear** [comment]
I was just thinking about this some more, and I wonder if we should handle the cookie problem more gracefully too.  I still think my suggestion to @joshr is the right thing to do but we should be able to deliver content to a Safari client even if cookies don't exist?

**Comment by JohnMcLear** [comment]
Here is a PR with a much better UX.  Safari / Brave users CAN use the pad even with x origin BUT if they refresh/rejoin Etherpad is not aware of who they are.

**Comment by joshr** [comment]
hey! great stuff, thanks for looking into this.

how about this turns into a plugin request? an "enable iframe embedding" plugin that etherpad hosts can use to let others embed their pads in their own sites?

**Comment by JohnMcLear** [comment]
A plugin won't fix the problem.  This is not an Etherpad problem.  It's a browser problem.  We can make it suck less but it will still suck if they use safari or brave.  Consider putting Etherpad in a subdomain please.

**Comment by joshr** [comment]
update: i've moved etherpad to a subdomain now and that does seem to have solved the safari issue. thanks John! i am still however getting a problem with the site seeming to get stuck loading (see image). it does this after a few minutes of use, possibly caused by being open in multiple tabs/windows or multiple users at once? not sure how to diagnose this :/

the "Loading...." overlay stays for ages, blocking the UI, and then you see this error after a few minutes
![image_2020-05-21_12-40-30](https://user-images.githubusercontent.com/36216/82558126-ea68e800-9b64-11ea-82b7-1e12ea65666d.png)

**Comment by JohnMcLear** [comment]
Oooh good progress, well done!  That overlay issue is horrible.  If you hit F12 do you get anything in developer console?  If so, copy paste it please :)

**Comment by joshr** [comment]
this is what i get while the Loading... overlay is there, before the error appears (takes a few minutes):

(index):1 Invalid 'X-Frame-Options' header encountered when loading 'https://handbook.reorganise.org/p/reorganisers-handbook': 'ALLOW-FROM https://reorganise.org' is not a recognized directive. The header will be ignored.
index.js:83 WebSocket connection to 'wss://handbook.reorganise.org/socket.io/?EIO=3&transport=websocket&sid=J1k7NrFU7yw8dbYlAAAg' failed: Error during WebSocket handshake: Unexpected response code: 400
r.doOpen @ index.js:83
pad.js?callback=require.define:25 Uncaught DOMException: Blocked a frame with origin "https://handbook.reorganise.org" from accessing a cross-origin frame.
    at Object.<anonymous> (https://handbook.reorganise.org/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define:25:1479)
    at t (https://handbook.reorganise.org/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define:25:388)
    at https://handbook.reorganise.org/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define:25:2187
    at Function.v.each.v.forEach (https://handbook.reorganise.org/javascripts/lib/ep_etherpad-lite/static/js/ace2_common.js?callback=require.define:23:1720)
    at Object.d.onEditorReady (https://handbook.reorganise.org/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define:25:2166)
    at readyFunc (https://handbook.reorganise.org/p/reorganisers-handbook:17:18)
    at eval (eval at R (https://handbook.reorganise.org/p/reorganisers-handbook:2:1418), <anonymous>:5:47536)

**Comment by JohnMcLear** [comment]
```
JQMIGRATE: Migrate is installed, version 1.4.1 jquery-migrate.min.js:2:552
Loading failed for the <script> with source “https://www.googletagmanager.com/gtag/js?id=UA-142176931-5”. reorganise.org:165:1
Invalid X-Frame-Options: “ALLOW-FROM https://reorganise.org” header from “https://handbook.reorganise.org/p/reorganisers-handbook” loaded into “https://reorganise.org/”.

Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user’s experience. For more help http://xhr.spec.whatwg.org/ require-kernel.js:1:1962
Couldn't find translation key pad.settings.about html10n.js:1:8258
Couldn't find translation key pad.settings.poweredBy html10n.js:1:8258
Firefox can’t establish a connection to the server at wss://handbook.reorganise.org/socket.io/?EIO=3&transport=websocket&sid=fWkeSoTlH3YHCE9dAAAk. socket.io.js:7:23187
Object { type: "CLIENT_VARS", data: {…} }
pad.js:3:5233
Object { type: "COLLABROOM", data: {…} }
pad.js:28:3448
Object { type: "COLLABROOM", data: {…} }
pad.js:28:3448
Object { type: "COLLABROOM", data: {…} }
pad.js:28:3448
Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user’s experience. For more help http://xhr.spec.whatwg.org/ reorganisers-handbook line 24 > scriptElement:13:1962
SecurityError: Permission denied to access property "console" on cross-origin object
```

**Comment by JohnMcLear** [comment]
Distills down to:
1. 
```
Invalid X-Frame-Options: “ALLOW-FROM https://reorganise.org” header from “https://handbook.reorganise.org/p/reorganisers-handbook” loaded into “https://reorganise.org/”.
```
Probably a misconfiguration of ep_helmet?

2.
```
SecurityError: Permission denied to access property "console" on cross-origin object
```
Something is console logging out to "top" one assumes, needs to be commented out, shouldn't be causing a problem though..


3. 
```
Firefox can’t establish a connection to the server at wss://handbook.reorganise.org/socket.io/?EIO=3&transport=websocket&sid=fWkeSoTlH3YHCE9dAAAk. socket.io.js:7:23187
```
Misconfiguration of nginx / apache config?  This will be a problem and needs addressing imho.

**Comment by joshr** [comment]
cool,

1. i only just installed that to see if it would fix this problem, followed the instructions.

2. no idea about that!

3. possibly, any clues? i have it setup to proxy the port at :80, no 404 rule (otherwise ep doesn't work at all), and http to https managed by certbot

here's my config:

server {

        root /var/www/handbook.reorganise.org/html;
        index index.html index.htm index.nginx-debian.html;

        server_name handbook.reorganise.org;

        location / {
                proxy_pass             http://localhost:9001/;
                proxy_set_header       Host $host;
                proxy_buffering off;
        }

    listen [::]:443 ssl; # managed by Certbot
    listen 443 ssl; # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/handbook.reorganise.org/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/handbook.reorganise.org/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot

}
server {
    if ($host = handbook.reorganise.org) {
        return 301 https://$host$request_uri;
    } # managed by Certbot


        listen 80;
        listen [::]:80;

        server_name handbook.reorganise.org;
    return 404; # managed by Certbot


}

**Comment by JohnMcLear** [comment]
1. Weird, for now I'm not sure we need it :)
2. Ignore it for now, if it causes a problem it's an easy fix.
3. I think more logic is reuqired for websockets :)  Afaik it's in the example config no?

**Comment by JohnMcLear** [comment]
proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection $connection_upgrade;

as per https://github.com/ether/etherpad-lite/wiki/How-to-put-Etherpad-Lite-behind-a-reverse-Proxy

**Comment by JohnMcLear** [comment]
1. see what happens if you remove ep_helmet btw, I think something is misconfigured in your settings blob.  Also if you can copy/paste that it'd be good :)  It's new functionality I didn't test in production yet.

**Comment by joshr** [comment]
here's the readout after the error shows btw:

```
(index):1 Invalid 'X-Frame-Options' header encountered when loading 'https://handbook.reorganise.org/p/reorganisers-handbook': 'ALLOW-FROM https://reorganise.org' is not a recognized directive. The header will be ignored.
index.js:83 WebSocket connection to 'wss://handbook.reorganise.org/socket.io/?EIO=3&transport=websocket&sid=DRD8BRoP1yCu7A4xAAAs' failed: Error during WebSocket handshake: Unexpected response code: 400
r.doOpen @ index.js:83
pad.js?callback=require.define:25 Uncaught DOMException: Blocked a frame with origin "https://handbook.reorganise.org" from accessing a cross-origin frame.
    at Object.<anonymous> (https://handbook.reorganise.org/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define:25:1479)
    at t (https://handbook.reorganise.org/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define:25:388)
    at https://handbook.reorganise.org/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define:25:2187
    at Function.v.each.v.forEach (https://handbook.reorganise.org/javascripts/lib/ep_etherpad-lite/static/js/ace2_common.js?callback=require.define:23:1720)
    at Object.d.onEditorReady (https://handbook.reorganise.org/javascripts/lib/ep_etherpad-lite/static/js/pad.js?callback=require.define:25:2166)
    at readyFunc (https://handbook.reorganise.org/p/reorganisers-handbook:17:18)
    at eval (eval at R (https://handbook.reorganise.org/p/reorganisers-handbook:2:1418), <anonymous>:5:47536)
timeslider.js?callback=require.define:22 Uncaught TypeError: Cannot read property 'length' of undefined
    at Object.toggleDropDown (timeslider.js?callback=require.define:22)
    at Object.showModal (timeslider.js?callback=require.define:26)
    at Object.disconnected (pad.js?callback=require.define:40)
    at r.<anonymous> (pad.js?callback=require.define:3)
    at r.emit (index.js:83)
    at r.onevent (index.js:83)
    at r.onpacket (index.js:83)
    at r.<anonymous> (index.js:83)
    at r.emit (index.js:83)
    at r.ondecoded (index.js:83)
toggleDropDown @ timeslider.js?callback=require.define:22
showModal @ timeslider.js?callback=require.define:26
disconnected @ pad.js?callback=require.define:40
(anonymous) @ pad.js?callback=require.define:3
r.emit @ index.js:83
r.onevent @ index.js:83
r.onpacket @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
r.ondecoded @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
a.add @ index.js:83
r.ondata @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
r.onPacket @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
r.onPacket @ index.js:83
n @ index.js:83
e.decodePayload @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
timeslider.js?callback=require.define:22 Uncaught TypeError: Cannot read property 'length' of undefined
    at Object.toggleDropDown (timeslider.js?callback=require.define:22)
    at Object.showModal (timeslider.js?callback=require.define:26)
    at Object.reconnecting (pad.js?callback=require.define:40)
    at Object.handleChannelStateChange [as onChannelStateChange] (pad.js?callback=require.define:3)
    at Object.O [as setChannelState] (pad.js?callback=require.define:28)
    at r.<anonymous> (pad.js?callback=require.define:3)
    at r.emit (index.js:83)
    at r.emit (index.js:83)
    at r.emitAll (index.js:83)
    at index.js:83
toggleDropDown @ timeslider.js?callback=require.define:22
showModal @ timeslider.js?callback=require.define:26
reconnecting @ pad.js?callback=require.define:40
handleChannelStateChange @ pad.js?callback=require.define:3
O @ pad.js?callback=require.define:28
(anonymous) @ pad.js?callback=require.define:3
r.emit @ index.js:83
r.emit @ index.js:83
r.emitAll @ index.js:83
(anonymous) @ index.js:83
setTimeout (async)
r.reconnect @ index.js:83
r.onclose @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
r.onClose @ index.js:83
(anonymous) @ index.js:83
setTimeout (async)
r.onHeartbeat @ index.js:83
(anonymous) @ index.js:83
setTimeout (async)
r.setPing @ index.js:83
r.onPacket @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
r.onPacket @ index.js:83
n @ index.js:83
e.decodePayload @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
XMLHttpRequest.send (async)
i.create @ index.js:83
i @ index.js:83
o.request @ index.js:83
o.doPoll @ index.js:83
r.poll @ index.js:83
r.onData @ index.js:83
(anonymous) @ index.js:83
r.emit @ index.js:83
i.onData @ index.js:83
i.onLoad @ index.js:83
hasXDR.n.onreadystatechange @ index.js:83
```

**Comment by JohnMcLear** [comment]
Just noise mate, gonna delete that comment :)

**Comment by joshr** [comment]
ok, looking at the sockets stuff.

here's the helmet settings:

```
"ep_helmet":{
  "csp":{
    "directives": {
      "objectSrc": ["'none'"],
      "upgradeInsecureRequests": true,
      "workerSrc": false
    }
  }
},

"ep_helmet": {
  "frameguard": {
    "action": "allow-from",
    "domain": "https://reorganise.org"
  }
},
```

**Comment by JohnMcLear** [comment]
ahaha I see what's wrong 🗡️ 

You can't have two blocks.  Your ep_helmet setings should just be

```
"ep_helmet": {
  "frameguard": {
    "action": "allow-from",
    "domain": "https://reorganise.org"
  }
},
```

**Comment by joshr** [comment]
ok removing that first block. the error was happening before this though, so might not solve.

**Comment by JohnMcLear** [comment]
```
app.use(frameguard({
  action: 'allow-from',
  domain: 'http://example.com'
}))
```
is the example from helmet so your config / the example looks good to me :)

**Comment by joshr** [comment]
> ```
>         proxy_set_header Upgrade $http_upgrade;
>         proxy_set_header Connection $connection_upgrade;
> ```
> 
> as per https://github.com/ether/etherpad-lite/wiki/How-to-put-Etherpad-Lite-behind-a-reverse-Proxy

adding these two lines has broken nginx :/

**Comment by JohnMcLear** [comment]
https://stackoverflow.com/questions/36011581/nginx-unknown-directive-proxy-pass

**Comment by JohnMcLear** [comment]
What error do you get from nginx when you add those lines?  
``tail /var/log/nginx.log``

**Comment by joshr** [comment]
```
Job for nginx.service failed because the control process exited with error code.
See "systemctl status nginx.service" and "journalctl -xe" for details.
```

**Comment by JohnMcLear** [comment]
yep follow that

**Comment by joshr** [comment]
from: systemctl status nginx.service

```
● nginx.service - A high performance web server and a reverse proxy server
   Loaded: loaded (/lib/systemd/system/nginx.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Thu 2020-05-21 12:58:04 UTC; 6min ago
     Docs: man:nginx(8)
  Process: 19282 ExecStop=/sbin/start-stop-daemon --quiet --stop --retry QUIT/5 --pidfile /run/nginx.pid (code=killed, signal=TERM)
  Process: 19327 ExecStartPre=/usr/sbin/nginx -t -q -g daemon on; master_process on; (code=exited, status=1/FAILURE)
 Main PID: 28774 (code=exited, status=0/SUCCESS)

```

doing this now https://stackoverflow.com/questions/36011581/nginx-unknown-directive-proxy-pass

**Comment by JohnMcLear** [comment]
As per: http://nginx.org/en/docs/http/websocket.html

```
    map $http_upgrade $connection_upgrade {
        default upgrade;
        ''      close;
    }
```

was missing in your config :)

**Comment by joshr** [comment]
ok then! cool, testing now

**Comment by JohnMcLear** [comment]
This is intriguing..  If I get error

```
Invalid X-Frame-Options: “ALLOW-FROM https://reorganise.org” header from “https://handbook.reorganise.org/p/reorganisers-handbook” loaded into “https://reorganise.org/”.
```
The pad loads and things work.

If I comment out ep_helmet config I am unable to load the pad when in iframe.  so ep_helmet is clearly working and it's clearly making a difference but it's not being accepted by browser or so!  

I'm gonna try and remove ep_helmet entirely and see what happens :)

**Comment by JohnMcLear** [comment]
1. Pad is loading properly <-- ws:// now fixed in nginx.
2. ep_helmet was increasing security but also this caused an error in console but not an error in practice.  <-- for now I removed ep_helmet
3. console potentially blocking page load?  <-- working on this now..

**Comment by JohnMcLear** [comment]
Interesting..

```
diff --git a/src/static/js/pad.js b/src/static/js/pad.js
index 622b57f3..839abc75 100644
--- a/src/static/js/pad.js
+++ b/src/static/js/pad.js
@@ -266,7 +266,7 @@ function handshake()
     else if (!receivedClientVars && obj.type == "CLIENT_VARS")
     {
       //log the message
-      if (window.console) console.log(obj);
+      // if (window.console) console.log(obj);

       receivedClientVars = true;

@@ -339,7 +339,7 @@ function handshake()
       if (obj.disconnect)
       {
         console.warn("FORCED TO DISCONNECT");
-        console.warn(obj);
+        // console.warn(obj);
         padconnectionstatus.disconnected(obj.disconnect);
         socket.disconnect();

```

Pads load fine now..  gonna put minify back on et al and see if it's okay..

**Comment by JohnMcLear** [comment]
Interesting, putting minify back on breaks it..  I am gonna isolate the console.log that's breaking things.

**Comment by JohnMcLear** [comment]
Okay minify back on and working.

I'm gonna nuke the top.console messages in:
``ace2_inner.js``
``ace.js``

I think as we tightened up security they were restricted..  Interesting..

**Comment by JohnMcLear** [comment]
https://github.com/ether/etherpad-lite/pull/4039
